### ISSUE TITLE
(proposal): Return a Date in UTC timezone from NgbDateNativeAdapter.toModel() method

### ISSUE THREAD

**Issue by AbakumovAlexandr** [ng-bootstrap/ng-bootstrap#2631]
### Proposal description:
Currently, when using `NgbDateNativeAdapter` and selecting a date, it would be passed to a model as a noon in a **browser time zone**:

> Sat Aug 03 2013 12:00:00 GMT-0500 (Central Daylight Time)

This creates a need to correct the time zone before posting it to a server, since most likely when server expects a `Date` (not `DateTime`), it doesn't like if the JSON value comes with a time zone other than UTC.

Since the most likely use case of `Datepicker` is to get a _date without time_ from a user and not to worry about its floating due to browser's timezone, it looks more appropriate to work with native JS `Date`s always in UTC in the case of `Datepicjer`.

### Version of Angular, ng-bootstrap, and Bootstrap:

ng-bootstrap: 3.0.0

**Comment by pkozlowski-opensource** [comment]
@maxokorokov will have more to say but my first knee-jerk reaction is that:
* we've introduced `NgbDataStruct` precisely to _not_ deal time-zone related issues - this is still our recommended data model;
* we've introduced model adapters so people can return whatever object they need - including a date object with a specific time-zone.

@maxokorokov WDYT? I've got an impression that we will never be able to cover all the needs on our end and we've got all the flex points in place, so I'm leaning towards "won't fix"...

**Comment by maxkorz** [comment]
@AbakumovAlexandr Here's a model adapter to convert internal NgbDateStruct to date in ISO format (the timezone is always zero UTC offset, as denoted by the suffix "Z".): https://stackblitz.com/edit/angular-c2kbbg?file=app%2Fngb-date-to-iso-string-adapter.ts

**Comment by AbakumovAlexandr** [comment]
@pkozlowski-opensource , sure, we'll never cover all the use cases, but I suppose the main use case of `Datepicker` is `to get **date** from a user and store it somehow, most likely, send it to server`.

In general, after days spent on dealing with dates in client-server apps, my main take-away is that a moment of date without a time part or, in other words, an amount of milliseconds since 1970 represented by the date must never depend on a user's timezone.

I, honestly, don't see any advantages of a `Date` returned from the component implied for getting dates without the time part being in the user time zone.

And, finally, convert from UTC to the user time zone (if one ever need it) is much simpler than vice versa, since UTC moment of time is always definitive and self-describing, while the user time zone is never the same and, basically, every developer will have to deal with the conversion.

**Comment by AbakumovAlexandr** [comment]
@pkozlowski-opensource and @maxokorokov , I know I constantly say it, but if the above doesn't sounds well for you, consider at least the approach taken in Angular Material, please :) Specifically, [this](https://material.angular.io/components/datepicker/overview#choosing-a-date-implementation-and-date-format-settings):

> By default the `MomentDateAdapter` will creates dates in your time zone specific locale. You can change the default behaviour to parse dates as UTC by providing the `MAT_MOMENT_DATE_ADAPTER_OPTIONS` and setting it to `useUtc: true`.
```
@NgModule({
  imports: [MatDatepickerModule, MatMomentDateModule],
  providers: [
    { provide: MAT_MOMENT_DATE_ADAPTER_OPTIONS, useValue: { useUtc: true } }
  ]
})
```

**Comment by AbakumovAlexandr** [comment]
@maxkorz , exactly! Make this as a default behavior:
new Date(Date.UTC(date.year, date.month - 1, date.day, 12))
instead of:
new Date(date.year, date.month - 1, date.day, 12)
in [`NgbDateNativeAdapter`](https://github.com/ng-bootstrap/ng-bootstrap/blob/88e5e984e1b2f530aaa36b2f70f55aef393d8163/src/datepicker/adapters/ngb-date-native-adapter.ts#L13) is what I'm proposing.

**Comment by AbakumovAlexandr** [comment]
@pkozlowski-opensource , BTW:

> we've introduced model adapters so people can return whatever object they need - including a date object with a specific time-zone.

Suppose a use case of some `Datepicker`s needed to assign `Date` in UTC time zone to a model, and others in browser time zone **within the same component**. So with the current design, there should be two model adapters, but we cannot provide them granular enough to achieve the desired result.

What's the recommended solution for use cases like this?

**Comment by AbakumovAlexandr** [comment]
@maxokorokov :

> somewhere in, say, Brazil and click on the MAY 25 in the datepicker → .UTC will return a date with MAY 25, but new Date will make it MAY 24 in the local timezone.

Right, However, what is the actual issue here? Why do your ever use a `Datepicker` and expect to get a date in a local time zone?

As I said, the very reason to use `Datepicker` is to get a date. The date shouldn't carry any time zone, because it does not have the time part by definition.

The huge advantage when dealing with `Datepicker` is to get a consistent JS `Date` object when selecting a consistent date regardless of user's time zones or any other variable. So, you always know that:

- Aug 24, 2018 is ALWAYS represented by 1535068800000
- Aug 24, 2018 is ALWAYS represented by 1535155200000
- etc.

When Aug 24, 2018 representation floats, that creates a zillion of potential bugs initially masked every app developer has to deal with. And the issue #2634 we're discussing in parallel with this one, is a bright example of this point.

**Comment by pkozlowski-opensource** [comment]
> As I said, the very reason to use Datepicker is to get a date. The date shouldn't carry any time zone, because it does not have the time part by definition.

Exactly! But we don't have such thing in JS, hence we've introduced `NgbDataStruct`. We still believe that using native JS dates is asking for troubles, regardless of how you massage it.

**Comment by maxkorz** [comment]
by the way, NgbDateISOAdapter was in [plans](https://github.com/ng-bootstrap/ng-bootstrap/issues/1977)

**Comment by AbakumovAlexandr** [comment]
@maxokorokov , does `NgbDateISOAdapter` imply that every single combination of year\month\day selected in `Datepicker` would produce a JS `Date` representing the exact same UNIX Epoch regardless of a user's timezone?

**Comment by pkozlowski-opensource** [comment]
`NgbDateISOAdapter` would produce a _string_ not a _Date_ object.

**Comment by AbakumovAlexandr** [comment]
@pkozlowski-opensource , right. Then `NgbDateISOAdapter` doesn't seem to be a solution if you want to bind a `Datepicker` to a `Date` model.
